### PR TITLE
fix: save auth state between pd sessions in secure storage

### LINKED ISSUE #39: Authentication state is showing "logged out" after restart of Podman Desktop
The extension claims to be logged out after a restart.  I'd love us to find a way to improve on that.  There may be ways to store/cache the refresh token or to check whether we're logged into `registry.redhat.io` and do a status check of subscription-manager inside the VM.  There may be other options as well.

Screenshot 2024-02-13 at 13 50 30

Ignore until https://github.com/redhat-developer/podman-desktop-redhat-account-ext/pull/35 has merged.

depends on https://github.com/containers/podman-desktop/issues/6044.

Cc: @dgolovin @benoitf @jeffmaury @slemeur

### PR DESCRIPTION
**Pull request by dgolovin**
Fix #39.

How to test:
1. run podman desktopn `yarn watch --extension-folder path/to/sso/ext
2. Login to redhat sso from Settings/Authentication
3. exit podman-desktop
4. run podman desktopn `yarn watch --extension-folder path/to/sso/ext
5. open Settings/Authentication, see provider shows status logged in

### CODE REVIEW

File: src/authentication-service.ts [hunk 1]
```diff
@@ -159,6 +157,13 @@ export class RedHatAuthenticationService {
         Logger.info('Failed to initialize stored data');
         await this.clearSessions();
       }
+
+      if (this.context.secrets) {
+        this._disposables.push(this.context.secrets.onDidChange(() => {
+          Logger.info('Secrets changed, checking for token updates');
+          this.checkForUpdates();
+        }));
+      }
```
Comment (dgolovin) [inline src/authentication-service.ts:166]:
```suggestion
      this._disposables.push(this.context.secrets.onDidChange(() => {
        Logger.info('Secrets changed, checking for token updates');
        this.checkForUpdates();
      }));
```

**Review by vrothberg — COMMENTED** [review]
Works like a charm, great work!

Let's assume a user would log out of the registry or re-create a podman machine.  Would the extension automatically re-login and re-subscribe (at start time)?

**Comment by vrothberg** [comment]
@benoitf WDYT?

**Comment by benoitf** [comment]
For podman machine I think that how it is done currently is not optimal. I mean, some work is done when you do 'login'. But even without exiting Podman Desktop, etc, people can delete or recreate podman machine at any time using cli or Podman Desktop, etc.

so the extension should monitor podman machines and create activation keys if not already done. I think it should not be tied only to the 'login' phase.

For the registry, if user logout from the registry then I would say it's the user's choice. Restarting Podman Desktop should still let the user unauthenticated.
Here, maybe in the UI of registries, if we know that we can signin to a registry without login/password we should have a way to tell : 'Login using Red Hat account'

File: src/authentication-service.ts [hunk 2]
```diff
@@ -159,6 +157,11 @@ export class RedHatAuthenticationService {
         Logger.info('Failed to initialize stored data');
         await this.clearSessions();
       }
+
+      this._disposables.push(this.context.secrets.onDidChange(() => {
+        Logger.info('Secrets changed, checking for token updates');
```
Comment (benoitf) [inline src/authentication-service.ts:162]: not sure about having that with info level or if it's needed ?

**Review by benoitf — APPROVED** [review]
unit tests would help there

**Comment by dgolovin** [comment]
@benoitf  @vrothberg  what if we would create extension that add web view the same way as AI studio does and we can name it the way it looks much better than 'Login to use Red Account Extension'? On that web view we can show status of VM subscription status and registry access status. There would be buttons to configure registry access and subscription activation.

File: src/authentication-service.ts (hunk 2 above)
Comment (dgolovin) [inline src/authentication-service.ts:162]:
```suggestion
```

**Comment by benoitf** [comment]
@dgolovin I think it would be a 'next step' improvement

**Comment by vrothberg** [comment]
Agreed.  We can tackle these questions later on.  This PR is a huge step in the right direction 🚀

**Comment by dgolovin** [comment]
> I think it would be a 'next step' improvement

I will open follow up issues for that.

**Comment by vrothberg** [comment]
Nice work!

**Pull request merged**